src/client.jl

```diff
@@ -233,52 +250,90 @@ function Base.iterate(iter::ChunkReader, _state=nothing)
+        input = Base.BufferStream()
+        write(input, body)
+        close(input)
+    else
+        input = nothing
+    end
+
     if stream
         @assert stream_to !== nothing
     end
 
     resp = nothing
+    output = Base.BufferStream()
 
-    function process_stream(http)
-        if body !== nothing
-            write(http, body)
-        end
-        resp = startread(http)
-        for chunk in ChunkReader(http)
-            return_type = ctx.client.get_return_type(ctx.return_type, String(copy(chunk)))
-            data = response(return_type, resp, chunk)
-            put!(stream_to, data)
+    if stream
+        @sync begin
+            @async begin
+                try
+                    resp = Downloads.request(resource_path;
+                        input=input,
+                        output=output,
+                        kwargs...
+                    )
+                    close(output)
+                catch ex
+                    @error("exception invoking request", exception=(ex,catch_backtrace()))
+                    rethrow()
+                end
+            end
+            @async begin
+                try
+                    for chunk in ChunkReader(output)
+                        return_type = ctx.client.get_return_type(ctx.return_type, String(copy(chunk)))
+                        data = response(return_type, resp, chunk)
+                        put!(stream_to, data)
+                    end
+                    close(stream_to)
+                catch ex
+                    @error("exception reading chunk", exception=(ex,catch_backtrace()))
+                    rethrow()
+                end
+            end
         end
+    else
+        resp = Downloads.request(resource_path;
+                    input=input,
+                    output=output,
+                    kwargs...
+                )
     end
 
-    if stream
-        HTTP.open(process_stream, uppercase(ctx.method), HTTP.URIs.URI(resource_path), ctx.header; kwargs...)
-        close(stream_to)
-    else
-        if body !== nothing
-            resp = HTTP.request(uppercase(ctx.method), HTTP.URIs.URI(resource_path), ctx.header, body; kwargs...)
-        else
-            resp = HTTP.request(uppercase(ctx.method), HTTP.URIs.URI(resource_path), ctx.header; kwargs...)
-        end
+    if ctx.curl_mime_upload !== nothing
+        LibCURL.curl_mime_free(ctx.curl_mime_upload)
+        ctx.curl_mime_upload = nothing
     end
 
-    return resp
+    return resp, output
 end
 
 function exec(ctx::Ctx, stream_to::Union{Channel,Nothing}=nothing)
     stream = stream_to !== nothing
-    resp = do_request(ctx, stream; stream_to=stream_to)
+    resp, output = do_request(ctx, stream; stream_to=stream_to)
 
     (200 <= resp.status <= 206) || throw(ApiException(resp))
 
-    return stream ? resp : response(ctx.client.get_return_type(ctx.return_type, resp), resp)
+    if stream
+        return resp
+    else
+        data = readavailable(output)
```

Comment:
Given that `output::BufferStream` and `output` is `close`d I think you can just `read(output)` here?

(Strangely! This close-then-read doesn't work at all with `IOBuffer`, which throws when it's closed!  For Julia 1.8 there's `closewrite` which would make more sense in these situations.)

Reply:
Yes I think Downloads never closes output?  But HTTP.jl always does.